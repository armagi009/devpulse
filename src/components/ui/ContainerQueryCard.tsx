'use client';

import React from 'react';
import { containerBreakpoints, fluidSpacing } from '@/lib/styles/responsive-layout';
import { containerQueryTypography } from '@/lib/styles/responsive-typography';

interface ContainerQueryCardProps {
  title: string;
  children: React.ReactNode;
  className?: string;
  containerName?: string;
  variant?: 'default' | 'compact' | 'expanded';
  headerActions?: React.ReactNode;
  footerActions?: React.ReactNode;
  image?: string;
}

/**
 * ContainerQueryCard Component
 * 
 * A card component that adapts its layout based on its container width
 * rather than the viewport width, using container queries.
 * 
 * This component demonstrates the power of container queries for component-level
 * responsive design, allowing components to adapt based on their parent container
 * size rather than the viewport size.
 */
const ContainerQueryCard: React.FC<ContainerQueryCardProps> = ({
  title,
  children,
  className = '',
  containerName,
  variant = 'default',
  headerActions,
  footerActions,
  image,
}) => {
  return (
    <div 
      className={`container-query-card ${className} ${variant}`}
      style={{
        containerType: 'inline-size',
        containerName: containerName || '',
        width: '100%',
        padding: '1rem',
        borderRadius: '0.5rem',
        boxShadow: '0 1px 3px rgba(0,0,0,0.12), 0 1px 2px rgba(0,0,0,0.24)',
        backgroundColor: 'white',
      }}
    >
      <style jsx>{`
        .card-content {
          display: flex;
          flex-direction: column;
        }
        
        .card-header {
          display: flex;
          flex-direction: column;
          margin-bottom: ${fluidSpacing.sm};
        }
        
        .card-title {
          font-size: ${containerQueryTypography.small.heading};
          font-weight: 600;
          margin-bottom: 0.5rem;
        }
        
        .card-body {
          font-size: ${containerQueryTypography.small.body};
        }
        
        .card-footer {
          margin-top: ${fluidSpacing.sm};
          display: flex;
          flex-direction: column;
        }
        
        .card-image {
          width: 100%;
          margin-bottom: ${fluidSpacing.sm};
          border-radius: 0.25rem;
          overflow: hidden;
        }
        
        .card-image img {
          width: 100%;
          height: auto;
          display: block;
        }
        
        /* Container query for medium-sized containers */
        @container (min-width: ${containerBreakpoints.medium}) {
          .card-content {
            flex-direction: row;
            flex-wrap: wrap;
          }
          
          .card-header {
            flex-direction: row;
            justify-content: space-between;
            align-items: center;
          }
          
          .card-title {
            font-size: ${containerQueryTypography.medium.heading};
            margin-bottom: 0;
            flex: 1;
          }
          
          .card-body {
            font-size: ${containerQueryTypography.medium.body};
          }
          
          .card-image {
            width: 30%;
            margin-right: ${fluidSpacing.md};
            margin-bottom: 0;
          }
          
          .card-footer {
            flex-direction: row;
            justify-content: flex-end;
          }
          
          /* Variant styles */
          .compact .card-content {
            flex-direction: column;
          }
          
          .expanded .card-content {
            flex-direction: column;
          }
          
          .expanded .card-image {
            width: 100%;
            margin-right: 0;
          }
        }
        
        /* Container query for large containers */
        @container (min-width: ${containerBreakpoints.large}) {
          .container-query-card {
            padding: 1.5rem;
          }
          
          .card-title {
            font-size: ${containerQueryTypography.large.heading};
          }
          
          .card-body {
            font-size: ${containerQueryTypography.large.body};
          }
          
          .card-image {
            width: 20%;
          }
          
          /* Variant styles */
          .expanded .card-content {
            flex-direction: row;
          }
          
          .expanded .card-image {
            width: 30%;
            margin-right: ${fluidSpacing.md};
          }
        }
      `}</style>
      
      <div className="card-content">
        {image && (
          <div className="card-image">
            <img src={image} alt={title} />
          </div>
        )}
        
        <div className="card-main">
          <div className="card-header">
            <h3 className="card-title">{title}</h3>
            {headerActions && <div className="card-actions">{headerActions}</div>}
          </div>
          
          <div className="card-body">
            {children}
          </div>
          
          {footerActions && (
            <div className="card-footer">
              {footerActions}
            </div>
          )}
        </div>
      </div>
    </div>
  );
};

export default ContainerQueryCard;