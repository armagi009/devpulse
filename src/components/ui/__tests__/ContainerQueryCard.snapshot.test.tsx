import React from 'react';
import { render } from '@testing-library/react';
import ContainerQueryCard from '../ContainerQueryCard';

// Mock the responsive-layout and responsive-typography utilities
jest.mock('@/lib/styles/responsive-layout', () => ({
  containerBreakpoints: {
    small: '20rem',
    medium: '40rem',
    large: '60rem',
  },
  fluidSpacing: {
    xs: '0.5rem',
    sm: '1rem',
    md: '1.5rem',
    lg: '2rem',
    xl: '3rem',
  },
}));

jest.mock('@/lib/styles/responsive-typography', () => ({
  containerQueryTypography: {
    small: {
      heading: '1.25rem',
      body: '0.875rem',
    },
    medium: {
      heading: '1.5rem',
      body: '1rem',
    },
    large: {
      heading: '1.75rem',
      body: '1.125rem',
    },
  },
}));

describe('ContainerQueryCard Snapshots', () => {
  it('renders default variant correctly', () => {
    const { container } = render(
      <ContainerQueryCard title="Default Card">
        <p>This is the card content</p>
      </ContainerQueryCard>
    );
    
    expect(container).toMatchSnapshot();
  });

  it('renders compact variant correctly', () => {
    const { container } = render(
      <ContainerQueryCard 
        title="Compact Card" 
        variant="compact"
      >
        <p>This is the compact card content</p>
      </ContainerQueryCard>
    );
    
    expect(container).toMatchSnapshot();
  });

  it('renders expanded variant correctly', () => {
    const { container } = render(
      <ContainerQueryCard 
        title="Expanded Card" 
        variant="expanded"
      >
        <p>This is the expanded card content</p>
      </ContainerQueryCard>
    );
    
    expect(container).toMatchSnapshot();
  });

  it('renders with image correctly', () => {
    const { container } = render(
      <ContainerQueryCard 
        title="Card with Image" 
        image="/test-image.jpg"
      >
        <p>This is a card with an image</p>
      </ContainerQueryCard>
    );
    
    expect(container).toMatchSnapshot();
  });

  it('renders with header actions correctly', () => {
    const { container } = render(
      <ContainerQueryCard 
        title="Card with Header Actions" 
        headerActions={<button>Action</button>}
      >
        <p>This is a card with header actions</p>
      </ContainerQueryCard>
    );
    
    expect(container).toMatchSnapshot();
  });

  it('renders with footer actions correctly', () => {
    const { container } = render(
      <ContainerQueryCard 
        title="Card with Footer Actions" 
        footerActions={<button>Submit</button>}
      >
        <p>This is a card with footer actions</p>
      </ContainerQueryCard>
    );
    
    expect(container).toMatchSnapshot();
  });

  it('renders with custom className correctly', () => {
    const { container } = render(
      <ContainerQueryCard 
        title="Card with Custom Class" 
        className="custom-class"
      >
        <p>This is a card with a custom class</p>
      </ContainerQueryCard>
    );
    
    expect(container).toMatchSnapshot();
  });

  it('renders with custom containerName correctly', () => {
    const { container } = render(
      <ContainerQueryCard 
        title="Card with Container Name" 
        containerName="custom-container"
      >
        <p>This is a card with a custom container name</p>
      </ContainerQueryCard>
    );
    
    expect(container).toMatchSnapshot();
  });
});