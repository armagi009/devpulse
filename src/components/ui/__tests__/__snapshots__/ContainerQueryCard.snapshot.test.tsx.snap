// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContainerQueryCard Snapshots renders compact variant correctly 1`] = `
<div>
  <div
    class="jsx-3957274593 container-query-card  compact"
    style="width: 100%; padding: 1rem; border-radius: 0.5rem; box-shadow: 0 1px 3px rgba(0,0,0,0.12), 0 1px 2px rgba(0,0,0,0.24); background-color: white;"
  >
    <div
      class="jsx-3957274593 card-content"
    >
      <div
        class="jsx-3957274593 card-main"
      >
        <div
          class="jsx-3957274593 card-header"
        >
          <h3
            class="jsx-3957274593 card-title"
          >
            Compact Card
          </h3>
        </div>
        <div
          class="jsx-3957274593 card-body"
        >
          <p>
            This is the compact card content
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ContainerQueryCard Snapshots renders default variant correctly 1`] = `
<div>
  <div
    class="jsx-3957274593 container-query-card  default"
    style="width: 100%; padding: 1rem; border-radius: 0.5rem; box-shadow: 0 1px 3px rgba(0,0,0,0.12), 0 1px 2px rgba(0,0,0,0.24); background-color: white;"
  >
    <div
      class="jsx-3957274593 card-content"
    >
      <div
        class="jsx-3957274593 card-main"
      >
        <div
          class="jsx-3957274593 card-header"
        >
          <h3
            class="jsx-3957274593 card-title"
          >
            Default Card
          </h3>
        </div>
        <div
          class="jsx-3957274593 card-body"
        >
          <p>
            This is the card content
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ContainerQueryCard Snapshots renders expanded variant correctly 1`] = `
<div>
  <div
    class="jsx-3957274593 container-query-card  expanded"
    style="width: 100%; padding: 1rem; border-radius: 0.5rem; box-shadow: 0 1px 3px rgba(0,0,0,0.12), 0 1px 2px rgba(0,0,0,0.24); background-color: white;"
  >
    <div
      class="jsx-3957274593 card-content"
    >
      <div
        class="jsx-3957274593 card-main"
      >
        <div
          class="jsx-3957274593 card-header"
        >
          <h3
            class="jsx-3957274593 card-title"
          >
            Expanded Card
          </h3>
        </div>
        <div
          class="jsx-3957274593 card-body"
        >
          <p>
            This is the expanded card content
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ContainerQueryCard Snapshots renders with custom className correctly 1`] = `
<div>
  <div
    class="jsx-3957274593 container-query-card custom-class default"
    style="width: 100%; padding: 1rem; border-radius: 0.5rem; box-shadow: 0 1px 3px rgba(0,0,0,0.12), 0 1px 2px rgba(0,0,0,0.24); background-color: white;"
  >
    <div
      class="jsx-3957274593 card-content"
    >
      <div
        class="jsx-3957274593 card-main"
      >
        <div
          class="jsx-3957274593 card-header"
        >
          <h3
            class="jsx-3957274593 card-title"
          >
            Card with Custom Class
          </h3>
        </div>
        <div
          class="jsx-3957274593 card-body"
        >
          <p>
            This is a card with a custom class
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ContainerQueryCard Snapshots renders with custom containerName correctly 1`] = `
<div>
  <div
    class="jsx-3957274593 container-query-card  default"
    style="width: 100%; padding: 1rem; border-radius: 0.5rem; box-shadow: 0 1px 3px rgba(0,0,0,0.12), 0 1px 2px rgba(0,0,0,0.24); background-color: white;"
  >
    <div
      class="jsx-3957274593 card-content"
    >
      <div
        class="jsx-3957274593 card-main"
      >
        <div
          class="jsx-3957274593 card-header"
        >
          <h3
            class="jsx-3957274593 card-title"
          >
            Card with Container Name
          </h3>
        </div>
        <div
          class="jsx-3957274593 card-body"
        >
          <p>
            This is a card with a custom container name
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ContainerQueryCard Snapshots renders with footer actions correctly 1`] = `
<div>
  <div
    class="jsx-3957274593 container-query-card  default"
    style="width: 100%; padding: 1rem; border-radius: 0.5rem; box-shadow: 0 1px 3px rgba(0,0,0,0.12), 0 1px 2px rgba(0,0,0,0.24); background-color: white;"
  >
    <div
      class="jsx-3957274593 card-content"
    >
      <div
        class="jsx-3957274593 card-main"
      >
        <div
          class="jsx-3957274593 card-header"
        >
          <h3
            class="jsx-3957274593 card-title"
          >
            Card with Footer Actions
          </h3>
        </div>
        <div
          class="jsx-3957274593 card-body"
        >
          <p>
            This is a card with footer actions
          </p>
        </div>
        <div
          class="jsx-3957274593 card-footer"
        >
          <button>
            Submit
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ContainerQueryCard Snapshots renders with header actions correctly 1`] = `
<div>
  <div
    class="jsx-3957274593 container-query-card  default"
    style="width: 100%; padding: 1rem; border-radius: 0.5rem; box-shadow: 0 1px 3px rgba(0,0,0,0.12), 0 1px 2px rgba(0,0,0,0.24); background-color: white;"
  >
    <div
      class="jsx-3957274593 card-content"
    >
      <div
        class="jsx-3957274593 card-main"
      >
        <div
          class="jsx-3957274593 card-header"
        >
          <h3
            class="jsx-3957274593 card-title"
          >
            Card with Header Actions
          </h3>
          <div
            class="jsx-3957274593 card-actions"
          >
            <button>
              Action
            </button>
          </div>
        </div>
        <div
          class="jsx-3957274593 card-body"
        >
          <p>
            This is a card with header actions
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ContainerQueryCard Snapshots renders with image correctly 1`] = `
<div>
  <div
    class="jsx-3957274593 container-query-card  default"
    style="width: 100%; padding: 1rem; border-radius: 0.5rem; box-shadow: 0 1px 3px rgba(0,0,0,0.12), 0 1px 2px rgba(0,0,0,0.24); background-color: white;"
  >
    <div
      class="jsx-3957274593 card-content"
    >
      <div
        class="jsx-3957274593 card-image"
      >
        <img
          alt="Card with Image"
          class="jsx-3957274593"
          src="/test-image.jpg"
        />
      </div>
      <div
        class="jsx-3957274593 card-main"
      >
        <div
          class="jsx-3957274593 card-header"
        >
          <h3
            class="jsx-3957274593 card-title"
          >
            Card with Image
          </h3>
        </div>
        <div
          class="jsx-3957274593 card-body"
        >
          <p>
            This is a card with an image
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
